# Assignment 1 - Probability, Linear Algebra, Programming, and Git

## *Vivek Sahukar*
Netid:  *vs165*

# Probability and Statistics Theory

## 1
Let $f(x) = \begin{cases}
                0           & x < 0  \\
                \alpha x^2  & 0 \leq x \leq 2 \\
                0           & 2 < x
            \end{cases}$
            
For what value of $\alpha$ is $f(x)$ a valid probability density function?

*Note: for all assignments, write out all equations and math for all assignments using markdown and [LaTeX](https://tobi.oetiker.ch/lshort/lshort.pdf) and show all work*

**ANSWER**

$$\int _{0}^{2}f\left( x\right) dx=1\\\int _{0}^{2}\alpha x^{2}dx=\alpha \left[ \begin{matrix} x^{3}\\ -\\ 3\end{matrix} \right] _{0}^{2}=\frac {\alpha } {3}\left[ \begin{matrix} 2^{3} - 0^{3}\end{matrix} \right] =\frac {8\alpha } {3}=1\\\Rightarrow \alpha =3 / 8 $$


## 2
What is the cumulative distribution function (CDF) that corresponds to the following probability distribution function? Please state the value of the CDF for all possible values of $x$.

$f(x) = \begin{cases}
    \frac{1}{3} & 0 < x < 3 \\
    0           & \text{otherwise}
    \end{cases}$

$CDF=\int _{-\infty }^{a}f\left( x\right) dx $

$CDF\left( x=1\right) =\int _{-\infty }^{1}f\left( x\right) dx=\int _{-\infty }^{0}0dx+\int _{0}^{1}\frac {2} {3}dx=0+\frac {1} {3}\left[ x\right] _{0}^{2}=\frac {1} {3}$

$CDF\left( x=2\right) =\int _{-\infty }^{2}f\left( x\right) dx=\int _{-\infty }^{0}0dx+\int _{0}^{2}\frac {1} {3}dx=0+\frac {1} {3}\left[ x\right] _{0}^{2}=\frac {2} {3}\$

$CDF\left( x=3\right) =\int _{-\infty }^{3}f\left( x\right) dx=\int _{-\infty }^{0}0dx+\int _{0}^{3}\frac {1} {3}dx=0+\frac {1} {3}\left[ x\right] _0^{3}=\frac {3} {3}=1$

## 3
For the probability distribution function for the random variable $X$,

$f(x) = \begin{cases}
    \frac{1}{3} & 0 < x < 3 \\
    0           & \text{otherwise}
    \end{cases}$
    
what is the (a) expected value and (b) variance of $X$. *Show all work*.

**ANSWER**

(a) Expected value:

$$E\left( x\right) =\int _{-\infty }^{\infty }xP\left( x\right) dx=\int _{-\infty }^{0}x\left( 0\right) dx+\int _{0}^{3}\frac {1} {3}xdx+\int _{3}^{\infty }x.0dx\\=0+\frac {1} {3}\left[ \begin{matrix} x^{2}\\ -\\ 2\end{matrix} \right] 0^{3}+0\\=\frac {3^{2}} {3.2}=\frac {3} {2} $$


(b) Variance:
$$V\left( x\right) =\int x^{2}f\left( x\right) dx-\mu ^{2}\\=\int _{-\infty }^{0}x^{2}0dx+\int _{0}^{3}\frac {1} {3}x^{2}dx+\int _{3}^{\infty }x^{2}0dx-\left( \frac {3} {2}\right) ^{2}\\=0+\frac {1} {3}\left[ \begin{matrix} x^{3}\\ -\\ 3\end{matrix} \right] _{0}^{3}+0-\frac {9} {4}\\=\frac {3^{3}} {3}-\frac {9} {4}=\frac {3} {4} $$


## 4
Consider the following table of data that provides the values of a discrete data vector $\mathbf{x}$ of samples from the random variable $X$, where each entry in $\mathbf{x}$ is given as $x_i$.

*Table 1. Dataset N=5 observations*

|        | $x_0$ | $x_1$ | $x_2$ | $x_3$ | $x_4$ |
|------  |-------|-------|-------|-------|-------|
|$\textbf{x}$| 2     | 3     | 10    | -1    | -1    |

What is the (a) mean, (b) variance, and the  of the data? 

*Show all work. Your answer should include the definition of mean, median, and variance in the context of discrete data.*

**ANSWER**

(a) Mean

$$Mean=E\left( x\right) =\sum _{i=1}^{k}x_{i}.P\left( X=x_{i}\right) =\frac {1} {5}\left[ 2+3+10-1-1\right] =\frac {13} {5}=2.6$$

(b) Variance

$$\sigma ^{2}=V\left( x\right) =\sum_{i=1}^{k}\left( x_{i}-\mu\right) ^{2}P\left( X=x_{i}\right) \\=\frac {1} {5}\left[ \left( 2-2.6\right) ^{2}+\left( 3-2.6\right) ^{2}+\left( 10-2.6\right) ^{2}+\left( -1-2.6\right) ^{2}+\left( -1-2.6\right) ^{2}\right] = 16.24 $$

## 5
Review of counting from probability theory. 

(a) How many different 7-place license plates are possible if the first 3 places only contain letters and the last 4 only contain numbers?

(b) How many different batting orders are possible for a baseball team with 9 players?

(c) How many batting orders of 5 players are possible for a team with 9 players total?

(d) Let's assume this class has 26 students and we want to form project teams. How many unique teams of 3 are possible?

*Hint: For each problem, determine if order matters, and if it should be calculated with or without replacement.*

**ANSWER**

(a) 

First 3 places can be chosen from 26 letters in ways: ${26 \choose 3}$ x 3! = 26.25.24

Last 4 places can be chosen from 10 digits in ways: ${10 \choose 4}$ x 4! = 10.9.8.7

Multiplying these two options gives the total number of ways = **78,624,000**

Order matters because changing the order of numbers or letters changes the license plate. Since, there is no condition about not repeating a certain number or alphabet, the choice is with replacement.

(b)

Different batting orders possible are 9! = **362,880**

Order matters and choice is without replacement since once a player has been chosen for a particular order, he/she cannot be chosen again.

(c)

First ways to choose 5 players out of 9 players = ${9 \choose 5}$

Then, those players can be permuted in 5! ways, since order matters.

Therefore, Different batting orders possible = ${9 \choose 5}$ x 5! = 15120

Order matters and choice is without replacement since once a player has been chosen for a particular order, he/she cannot be chosen again.

(d)

Number of ways to form unique teams of 3 students = ${26 \choose 3} = 2600

Here, order does not matter, since team should be unique. Moreover, the choice is without replacement since once a student has been cho



# Linear Algebra

## 6
**Matrix manipulations and multiplication**. Machine learning involves working with many matrices, so this exercise will provide you with the opportunity to practice those skills.

Let
$\mathbf{A} =  \begin{bmatrix}
1 & 2 & 3 \\
2 & 4 & 5 \\
3 & 5 & 6 
\end{bmatrix}$, $\mathbf{b} =  \begin{bmatrix}
-1  \\
3  \\
8  
\end{bmatrix}$, $\mathbf{c} =  \begin{bmatrix}
4  \\
-3  \\
6  
\end{bmatrix}$, and $\mathbf{I} =  \begin{bmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1 
\end{bmatrix}$

Compute the following or indicate that it cannot be computed:

1. $\mathbf{A}\mathbf{A}$
2. $\mathbf{A}\mathbf{A}^T$
3. $\mathbf{A}\mathbf{b}$
4. $\mathbf{A}\mathbf{b}^T$
5. $\mathbf{b}\mathbf{A}$
6. $\mathbf{b}^T\mathbf{A}$
7. $\mathbf{b}\mathbf{b}$
8. $\mathbf{b}^T\mathbf{b}$
9. $\mathbf{b}\mathbf{b}^T$
10. $\mathbf{b} + \mathbf{c}^T$
11. $\mathbf{b}^T\mathbf{b}^T$
12. $\mathbf{A}^{-1}\mathbf{b}$
13. $\mathbf{A}\circ\mathbf{A}$
14. $\mathbf{b}\circ\mathbf{c}$

*Note: The element-wise (or Hadamard) product is the product of each element in one matrix with the corresponding element in another matrix, and is represented by the symbol "$\circ$".*

In [17]:
import numpy as np
from numpy import linalg as LA

# wherever the multiplication is not possible due to dimensions mismatch between 2 matrices, Value error is printed.

A = np.array([[1,2,3],
             [2,4,5],
             [3,5,6]
             ])

b = np.array([-1, 3, 8]).reshape(3, 1)

c = np.array([4, -3, 6]).reshape(3, 1)

I = np.identity(3)


In [18]:
# ANSWER
# 1. AA
mat1 = A @ A
mat2 = A @ A.T
mat3 = A @ b
# mat4 (A @ b.T) not possible due to dimensions mismatch: shapes (3,3) and (1,3) not aligned: 3 (dim 1) != 1 (dim 0)
# mat5 (b @ A) not possible due to dimensions mismatch: shapes (3,1) and (3,3) not aligned: 1 (dim 1) != 3 (dim 0)
mat6 = b.T @ A
# mat7 (b @ b) not possible due to dimensions mismatch: shapes (3,1) and (3,1) not aligned: 1 (dim 1) != 3 (dim 0)
mat8 = b.T @ b
mat9 = b @ b.T
mat10 = b + c.T
# Python recycles the vector hence the sum (3x3 matrix) is produced.
# However, mat10 is not possible since the dimensions of b(3,1) and c.transpose(1,3) don't match. 
# mat11 (b.T @ b.T) is not possible due to dimensions mismatch: shapes (1,3) and (1,3) not aligned: 3 (dim 1) != 1 (dim 0)
mat12 = LA.inv(A) @ b
mat13 = np.multiply(A, A)
mat14 = np.multiply(b, c)

$\mathbf{matrix1} = 
\begin{bmatrix}
   14.0 &   25.0 &   31.0\\
   25.0 &   45.0 &   56.0\\
   31.0 &   56.0 &   70.0
\end{bmatrix}$

$\mathbf{matrix2} =
\begin{bmatrix}
   14.0 &   25.0 &   31.0\\
   25.0 &   45.0 &   56.0\\
   31.0 &   56.0 &   70.0
\end{bmatrix}$

$\mathbf{matrix3} = 
\begin{bmatrix}
   29.0\\
   50.0\\
   60.0
\end{bmatrix}$

$\mathbf{matrix4}$ is not possible due to dimensions mismatch: shapes (3,3) and (1,3) not aligned: 3 (dim 1) != 1 (dim 0)

$\mathbf{matrix5}$ is not possible due to dimensions mismatch: shapes (3,1) and (3,3) not aligned: 1 (dim 1) != 3 (dim 0)

$\mathbf{matrix6} =
\begin{bmatrix}
   29.0 &   50.0 &   60.0
\end{bmatrix}$

$\mathbf{matrix7}$ is not possible due to dimensions mismatch: shapes (3,1) and (3,1) not aligned: 1 (dim 1) != 3 (dim 0)

$\mathbf{matrix8} =
\begin{bmatrix}
   74.0
\end{bmatrix}$

$\mathbf{matrix9} =
\begin{bmatrix}
    1.0 &  -3.0 &  -8.0\\
  -3.0 &    9.0 &   24.0\\
  -8.0 &   24.0 &   64.0
\end{bmatrix}$

$\mathbf{matrix10} =
\begin{bmatrix}
    3.0 &  -4.0 &    5.0\\
    7.0 &    0.0 &    9.0\\
   12.0 &    5.0 &   14.0
\end{bmatrix}$

*Python recycles the vector hence the sum (3x3 matrix) is produced.*
**However, matrix10 is not possible since the dimensions of b(3,1) and c.transpose(1,3) don't match.**

$\mathbf{matrix11}$ is not possible due to dimensions mismatch: shapes (1,3) and (1,3) not aligned: 3 (dim 1) != 1 (dim 0)

$\mathbf{matrix12} =
\begin{bmatrix}
    6.0\\
    4.0\\
  -5.0
\end{bmatrix}$

$\mathbf{matrix13} =
\begin{bmatrix}
    1.0 &    4.0 &    9.0\\
    4.0 &   16.0 &   25.0\\
    9.0 &   25.0 &   36.0
\end{bmatrix}$

$\mathbf{matrix14} =
\begin{bmatrix}
  -4.0\\
  -9.0\\
   48.0
\end{bmatrix}$

## 6
**Eigenvectors and eigenvalues**. Eigenvectors and eigenvalues are useful for some machine learning algorithms, but the concepts take time to solidly grasp. For an intuitive review of these concepts, explore this [interactive website at Setosa.io](http://setosa.io/ev/eigenvectors-and-eigenvalues/). Also, the series of linear algebra videos by Grant Sanderson of 3Brown1Blue are excellent and can be viewed on youtube [here](https://www.youtube.com/playlist?list=PLZHQObOWTQDPD3MizzM2xVFitgF8hE_ab).

1. Calculate the eigenvalues and corresponding eigenvectors of matrix $\mathbf{A}$ above, from the last question.
2. Choose one of the eigenvector/eigenvalue pairs, $\mathbf{v}$ and $\lambda$, and show that $\mathbf{A} \mathbf{v} = \lambda \mathbf{v}$. Also show that this relationship extends to higher orders: $\mathbf{A} \mathbf{A} \mathbf{v} = \lambda^2 \mathbf{v}$
3. Show that the eigenvectors are orthogonal to one another (e.g. their inner product is zero). This is true for real, symmetric matrices.

In [19]:
# 6.1
w, v = LA.eig(A)

# 6.2

w1 = w[1] # extract second eigenvalue
v1 = v[:, 1] # extract second eigenvector corresponding to second eigenvalue
A @ v1.reshape(3,1) # multiply matrix by eigenvector
np.dot(A, v1) # another method to multiply matrix by eigenvector
w1 * v1 # multiply eigenvalue with eigenvector

# showing that his relationship extends to higher orders: AAv = lambda^2 v
AAv = A @ (np.dot(A, v1))
w1v1 = (w1**2) * v1

# 6.3 
v0, v1, v2 = [v[:, i] for i in range(3)] # Extract each column of eigenmatrix as eigenvector
v0v1 = np.dot(v0, v1)
v1v2 = np.dot(v1, v2)
v2v0 = np.dot(v2, v0)
# all values are extremely small so the inner product of the eigenvectors is zero.Thus the eignvectors are orthogonal.


**ANSWER**

### 6.1
$\mathbf{eigenvalue1, \lambda_{0}} = 11.3448$, $\mathbf{eigenvector1, v_{0}} = 
\begin{bmatrix}
 -0.32798 \\ -0.59100 \\ -0.73697
\end{bmatrix}$

$\mathbf{eigenvalue2, \lambda_{1}} = -0.5157$, $\mathbf{eigenvector2, v_{1}} = 
\begin{bmatrix}
 -0.73697 \\ -0.32798 \\ -0.59100
\end{bmatrix}$

$\mathbf{eigenvalue3, \lambda_{2}} = 0.1709$, $\mathbf{eigenvector3, v_{2}} = 
\begin{bmatrix}
 0.59100 \\ -0.73697 \\ 0.32798
\end{bmatrix}$

### 6.2

$\mathbf{A}\mathbf{v_{1}}$ = $\begin{bmatrix}
  0.38008\\
  0.16915\\
 -0.30480
\end{bmatrix}$, $\lambda_{2} \mathbf{v}$ = $\begin{bmatrix}
  0.38008 \\  0.16915 \\ -0.30480
\end{bmatrix}$

Hence, $\mathbf{A}\mathbf{v_{1}} = \lambda_{2} \mathbf{v}$

Also, $\mathbf{A} \mathbf{A} \mathbf{v_{1}}$ = $\begin{bmatrix}
 -0.19601 \\ -0.08723 \\  0.15719
\end{bmatrix}$, $\lambda_2^2 \mathbf{v}$ = $\begin{bmatrix}
 -0.19601 \\ -0.08723 \\  0.15719
\end{bmatrix}$


Hence, $\mathbf{A} \mathbf{A} \mathbf{v_{1}} = \lambda_2^2 \mathbf{v}$

### 6.3

$\mathbf{v_{0}} . \mathbf{v_{1}}$ = -2.22e-16; 
$\mathbf{v_{1}} . \mathbf{v_{2}}$ = -1.05e-15; 
$\mathbf{v_{2}} . \mathbf{v_{0}}$ = -4.44e-16

All dot product values are extremely small so the inner product of the eigenvectors is zero. Thus the eignvectors are orthogonal.

# Numerical Programming

## 7
Speed comparison between vectorized and non-vectorized code. Begin by creating an array of 10 million random numbers using the numpy random.randn module. Compute the sum of the squares first in a for loop, then using Numpy's `dot` module. Time how long it takes to compute each and report the results and report the output. How many times faster is the vectorized code than the for loop approach?

*Note: all code should be well commented, properly formatted, and your answers should be output using the `print()` function as follows (where the # represents your answers, to a reasonable precision):

`Time [sec] (non-vectorized): ######`

`Time [sec] (vectorized):     ######`

`The vectorized code is ##### times faster than the vectorized code`

**ANSWER**

In [13]:
import numpy as np
import time
np.random.seed(1) # setting the random seed 

# Generate the random samples
s = np.random.randn(10000000)

# Compute the sum of squares the non-vectorized way (using a for loop)
start_time_long = time.time()
sum_of_squares_non_vectorized = 0
for i in s:
    sum_of_squares_non_vectorized += i**2
end_time_long = time.time()
time_non_vectorized = end_time_long - start_time_long

# Compute the sum of squares the vectorized way (using numpy)
start_time_short = time.time()
sum_of_squares_vectorized = np.dot(s, s)
end_time_short = time.time()
time_vectorized = end_time_short - start_time_short

times_comparison = time_non_vectorized / time_vectorized

# Print the results
print(f'Time [sec] (non-vectorized): {time_non_vectorized}\n')
print(f'Time [sec] (vectorized): {time_vectorized}\n')
print(f'The vectorized code is {times_comparison} times faster than non-vectorized code')

## 8
One popular Agile development framework is Scrum (a paradigm recommended for data science projects). It emphasizes the continual evolution of code for projects, becoming progressively better, but starting with a quickly developed minimum viable product. This often means that code written early on is not optimized, and that's a good thing - it's best to get it to work first before optimizing. Imagine that you wrote the following code during a sprint towards getting an end-to-end system working. Vectorize the following code and show the difference in speed between the current implementation and a vectorized version.

The function below computes the function $f(x,y) = x^2 - 2 y^2$ and determines whether this quantity is above or below a given threshold, `thresh=0`. This is done for $x,y \in \{-4,4\}$, over a 2,000-by-2,000 grid covering that domain.

(a) Vectorize this code and demonstrate (as in the last exercise) the speed increase through vectorization and (b) plot the resulting data - both the function $f(x,y)$ and the thresholded output - using [`imshow`](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.imshow.html?highlight=matplotlib%20pyplot%20imshow#matplotlib.pyplot.imshow) from `matplotlib`.

*Hint: look at the `numpy` [`meshgrid`](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.meshgrid.html) documentation*

In [14]:
import numpy as np
import time
import matplotlib.pyplot as plt

# Initialize variables for this exerise
nvalues = 2000
xvalues = np.linspace(-4,4,nvalues)
yvalues = np.linspace(-4,4,nvalues)
thresh  = 0



# Nonvectorized implementation

# Vectorized implementation
X,Y = np.meshgrid(xvalues,yvalues)
F = X**2 - 2 * Y**2
F_threshold = F > thresh

# Print the time for each and the speed increase

# Plot the result
# Plot the function f(x,y)
plot1 = plt.imshow(F, extent=(-4,4,-4,4), cmap='RdBu')
plt.xlabel('x')
plt.ylabel('y')
plt.title('$f(x,y) = x^2 - 2 * y^2$')
color_bar_max = max([F.max(), abs(F.min())])
plt.gcf().colorbar(plot1, label='$f(x,y)$')
plot1.set_clim(-color_bar_max,color_bar_max)
plt.show()

# Plot the thresholded output f(x,y) > 0
plot2 = plt.imshow(F_threshold, extent=(-4,4,-4,4), cmap='gray')
plt.xlabel('x')
plt.ylabel('y')
plt.title('$f(x,y) > 0$')
plt.gcf().colorbar(plot2, label='$f(x,y) > 0$')
plt.show()

## 9
This exercise will walk through some basic numerical programming exercises.
1. Synthesize $n=10^4$ normally distributed data points with mean $\mu=2$ and a standard deviation of $\sigma=1$. Call these observations from a random variable $X$, and call the vector of observations that you generate, $\textbf{x}$.
2. Calculate the mean and standard deviation of $\textbf{x}$ to validate (1) and provide the result to a precision of four significant figures.
3. Plot a histogram of the data in $\textbf{x}$ with 30 bins
4. What is the 90th percentile of $\textbf{x}$? The 90th percentile is the value below which 90% of observations can be found.
5. What is the 99th percentile of $\textbf{x}$?
6. Now synthesize $n=10^4$ normally distributed data points with mean $\mu=0$ and a standard deviation of $\sigma=3$. Call these observations from a random variable $Y$, and call the vector of observations that you generate, $\textbf{y}$.
7. Plot the histogram of the data in $\textbf{y}$ on a (new) plot with the histogram of $\textbf{x}$, so that both histograms can be seen and compared.
8. Using the observations from $\textbf{x}$ and $\textbf{y}$, estimate $E[XY]$

**ANSWER**

In [21]:
import numpy as np
import matplotlib.pyplot as plt
np.random.seed(1) 

# 1. 
print('1.)')
x = np.random.randn(10000) + 2
print('Examples from x: {}'.format(x[0:5]))

# 2
print('x_mean = {:.4}'.format(x.mean()))
print('x_std  = {:.4}'.format(x.std()))

# 3
plt.hist(x,bins=30)
plt.xlabel('x')
plt.ylabel('Number of Samples')
plt.grid('on')
plt.show()

# 4
percentile90 = np.percentile(x,90)
print('90th percentile = {:.4}'.format(percentile90))

# 5
percentile99 = np.percentile(x,99)
print('90th percentile = {:.4}'.format(percentile99))

# 6
y = 3*np.random.randn(10000)
print('Examples from y: {}'.format(y[0:5]))

# 7
histx, binsx = np.histogram(x, bins=30)
histy, binsy = np.histogram(y, bins=30)
centerx = (binsx[:-1] + binsx[1:]) / 2
centery = (binsy[:-1] + binsy[1:]) / 2
plt.plot(centerx,histx,'r-',label='x')
plt.plot(centery,histx,'b-',label='y')
plt.xlabel('Values')
plt.ylabel('Number of samples')
plt.grid('on')
plt.show()

# 8


ImportError: cannot import name 'get_backend' from 'matplotlib' (/Users/vs/anaconda3/lib/python3.7/site-packages/matplotlib/__init__.py)

## 10
Estimate the integral of the function $f(x)$ on the interval $0\leq x < 2.5$ assuming we only know the following points from $f$:

*Table 1. Dataset containing n=5 observations*

| $x_i$ | 0.0 | 0.5 | 1.0 | 1.5 | 2.0 |
|-|-|-|-|-|-|
| $y_i$ | 6 | 7 | 8 | 4 | 1 |


In [20]:
x = np.array([0, 0.5, 1.0, 1.5, 2.0])
y = np.array([6, 7, 8, 4, 1])

y_dif = (y[:-1] + y[1:]) / 2
dx = x[1:] - x[:-1]

int_y = sum(y_dif * dx )

print(f'The integral of y is: {int_y}')

The integral of y is: 11.25


**ANSWER**
11.25


# Version Control via Git

## 11
Complete the [Atlassian Git tutorial](https://www.atlassian.com/git/tutorials/what-is-version-control), specifically the following sections. Try each concept that's presented. For this tutorial, instead of using BitBucket, use Github. Create a github account here if you don't already have one: https://github.com/
1. [What is version control](https://www.atlassian.com/git/tutorials/what-is-version-control)
2. [What is Git](https://www.atlassian.com/git/tutorials/what-is-git)
3. [Install Git](https://www.atlassian.com/git/tutorials/install-git)
4. [Setting up a repository](https://www.atlassian.com/git/tutorials/install-git)
5. [Saving changes](https://www.atlassian.com/git/tutorials/saving-changes)
6. [Inspecting a repository](https://www.atlassian.com/git/tutorials/inspecting-a-repository)
7. [Undoing changes](https://www.atlassian.com/git/tutorials/undoing-changes)
8. [Rewriting history](https://www.atlassian.com/git/tutorials/rewriting-history)
9. [Syncing](https://www.atlassian.com/git/tutorials/syncing)
10. [Making a pull request](https://www.atlassian.com/git/tutorials/making-a-pull-request)
11. [Using branches](https://www.atlassian.com/git/tutorials/using-branches)
12. [Comparing workflows](https://www.atlassian.com/git/tutorials/comparing-workflows)

For your answer, affirm that you either completed the tutorial or have previous experience with all of the concepts above. Do this by typing your name below and selecting the situation that applies from the two options in brackets.

**ANSWER**

*I, [**Vivek Sahukar**], affirm that I have [**completed the above tutorial**]*

## 12
Using Github to create a static HTML website:
1. Create a branch in your `machine-learning-course` repo called "gh-pages" and checkout that branch (this will provide an example of how to create a simple static website using [Github Pages](https://pages.github.com/))
2. Create a file called "index.html" with the contents "Hello World" and add, commit, and push it to that branch.
3. Submit the following: (a) a link to your github repository and (b) a link to your new "Hello World" website. The latter should be at the address https://[USERNAME].github.io/ECE590-assignment0 (where [USERNAME] is your github username).

**ANSWER**



# Exploratory Data Analysis
## 13
Here you'll bring together some of the individual skills that you demonstrated above and create a Jupyter notebook based blog post on data analysis.

1. Find a dataset that interests you and relates to a question or problem that you find intriguing
2. Using a Jupyter notebook, describe the dataset, the source of the data, and the reason the dataset was of interest.
3. Check the data and see if they need to be cleaned: are there missing values? Are there clearly erroneous values? Do two tables need to be merged together? Clean the data so it can be visualized.
3. Plot the data, demonstrating interesting features that you discover. Are there any relationships between variables that were surprising or patterns that emerged? Please exercise creativity and curiosity in your plots.
4. What insights are you able to take away from exploring the data? Is there a reason why analyzing the dataset you chose is particularly interesting or important? Summarize this as if your target audience was the readership of a major news organization - boil down your findings in a way that is accessible, but still accurate.
5. Create a public repository on your github account titled "machine-learning-course". In it, create a readme file that contains the heading "ECE590: Introductory Machine Learning for Data Science". Add, commit, and push that Jupyter notebook to the master branch. Provide the link to the that post here.

**ANSWER**

